## 1.Java异常架构与异常关键字

### 1.1 Java异常简介

java异常是Java提供的一种识别及响应错误的机制

1.2 Java异常架构

![image-20210815081613600](https://kangrui-pictures.oss-cn-beijing.aliyuncs.com/img/image-20210815081613600.png)

#### Throwable

这是所有**错误**和**异常**的超类

#### Error（错误）

定义：Error类及其子类，程序中无法处理的错误。

特点：此类错误一般表示代码运行时JVM的问题，通常有 Virtual MachineError（虚拟机运行错误）、NoClassDefFoundError（类定义错误）等。比如 OutOfMemoryError：内存不足错误；StackOverflowError：栈溢出错误。此类错误发生时，JVM 将终止线程。

#### Exception（异常）

程序本身可以捕获并且可以处理的异常。Exception又分为运行时异常和编译时异常。

**运行时异常**

定义：RuntimeException类及其子类，表示JVM在运行时期间可能出现的异常

特点：编译器不会去检查它，也就是，当程序中可能出现的这类异常，倘若没有”通过throws声明抛出它“，也“没有try-catch语句捕获异常”，他还是会编译通过，比如空指针异常NullPointerException、数组下角标越界ArrayIndexOutBoundException、类转换异常ClassCastException、ArithemmeticException算数异常。**在程序中可以选择捕获处理，也可以不处理。**虽然 Java 编译器不会检查运行时异常，但是我们也可以通过 throws 进行声明抛出，也可以通过 try-catch 对它进行捕获处理。如果产生运行时异常，则需要通过修改代码来进行避免。例如，若会发生除数为零的情况，则需要通过代码避免该情况的发生！

RuntimeException异常会由Java虚拟机自动抛出自动捕获，此类异常大多数情况是代码本身有问题，应该从逻辑上解决改进代码。

**编译时异常**

定义：Exception中除了RuntimeException及其子类之外的异常

特点：Java编译器会检查它，如果程序中出现此类异常，比如：没有找到指定的类异常ClassNotFoundException、IO异常IOException。要么通过throws进行声明抛出，要么通过try-catch进行捕获处理，否则不能通过编译。在程序中，通常不会自定义该类异常，而是直接使用系统提供的异常类。**该异常我们必须手动在代码里添加捕获语句来处理该异常**。

#### 受检异常和非受检异常

Java所有的异常可以分为受检异常和非受检异常

**受检异常**

除 RuntimeException 及其子类外，其他的 Exception 异常都属于受检异常。编译器会检查此类异常，也就是说当编译器检查到应用中的某处可能会此类异常时，将会提示你处理本异常——要么使用try-catch捕获，要么使用方法签名中用 throws 关键字抛出，否则编译不通过。
**非受检异常**

编译器不会进行检查并且不要求必须处理的异常，也就说当程序中出现此类异常时，即使我们没有try-catch捕获它，也没有使用throws抛出该异常，编译也会正常通过。**该类异常包括运行时异常（RuntimeException极其子类）和错误（Error）。**

### 1.2 Java异常的关键字

try-用于监听，将要被监听的代码(可能抛出异常的代码)放在try语句块之内，当try语句块内发生异常时，异常就被抛出。

catch-用于捕获，catch捕获try语句的异常

finally-语句块总是会被执行

throw-用于抛出异常

throws 用于在方法签名中，用于该方法可能抛出的异常

## 2.Java异常处理

![image-20210815090531935](https://kangrui-pictures.oss-cn-beijing.aliyuncs.com/img/image-20210815090531935.png)

### 2.1 声明异常

通常，应该捕获那些知道如何处理的异常，将不知道如何处理的异常继续传递下去，传递异常可以在方法签名处使用throws关键字声明可能会抛出的异常

特点

非检查异常（Error和RuntimeException或他们子类）不可以使用throws关键字声明要抛出的异常

一个方法出现编译时异常，需要try-catch/throws处理，否则编译不通过

### 2.2 抛出异常

如果你觉得解决不了某些异常问题，且不需要调用者处理，那么你可以抛出异常。

throw关键字作用是在方法内部抛出一个`Throwable`类型的异常。**任何Java代码都可以通过throw语句抛出异常。**

### 2.3 捕获异常

程序运行之前不会报错，但是运行之后可能会出现某些未知的错误，但是还不想抛出到上一级，就需要通过try-catch来捕获异常。

### 2.4 如何选择异常类型

![image-20210815091422076](https://kangrui-pictures.oss-cn-beijing.aliyuncs.com/img/image-20210815091422076.png)

### 2.5 常见的异常处理

#### 直接抛出异常

通常，应该捕获那些知道如何处理的异常，将不知道如何处理的异常继续传递下去。传递异常可以在方法签名处使用 **throws** 关键字声明可能会抛出的异常。

```java
private static void readFile(String filePath) throws IOException {
    File file = new File(filePath);
    String result;
    BufferedReader reader = new BufferedReader(new FileReader(file));
    while((result = reader.readLine())!=null) {
        System.out.println(result);
    }
    reader.close();
}
```

#### 封装异常在抛出

有时我们会从 catch 中抛出一个异常，目的是为了改变异常的类型。多用于在多系统集成时，当某个子系统故障，异常类型可能有多种，可以用统一的异常类型向外暴露，不需暴露太多内部异常细节。

```java
private static void readFile(String filePath) throws MyException {    
    try {
        // code
    } catch (IOException e) {
        MyException ex = new MyException("read file failed.");
        ex.initCause(e);
        throw ex;
    }
}
```

## 3.Java异常常见面试题

### 3.1 Error和Exception的区别是什么

Error类型的错误通常称为虚拟机错误，如系统崩溃，内存不足，堆栈溢出。编译不会对这类错误检查，Java程序也不会对这类错误捕获，一旦发生了Error，程序就会终止，仅靠程序本身无法恢复。

Exception类错误可以在程序中进行捕获，遇到这类错误，应对其进行处理，程序便可以继续运行。

### 3.2 运行时异常和一般异常（受检异常）有什么区别

运行时异常包括RuntimeException及其子类，表示在JVM运行期间出现的异常，Java编译器不会检查。

受检异常是Exception中除了RuntimeException及其子类之外的异常，Java编译器会检测这些异常。

二者区别：强调调用者是否一定要处理此类异常，如果必须处理，那就是用受检异常，否则就使用非受检异常。

### 3.3 JVM如何处理异常

在一个方法出现异常之后，这个方法会创建一个异常对象，转交给JVM，该异常对象包括异常名称、异常描述以及异常发生时程序的状态。

创建异常对象转交给JVM的过程称为抛出异常。可能有一系列方法调用，最终才进入抛出异常的方法，这一系列方法调用的有序列表叫做调用栈。

JVM会顺着调用栈去查看是否有可以处理异常的代码块，如果有，则调用处理异常代码块。

### 3.4 throw和throws区别

java中的异常处理除了捕获异常和处理异常。还包括设声明异常和抛出异常。

声明异常，throws，在方法名上声明该关键字要跑出的异常。

抛出异常，throw，在方法内部通过throw抛出异常对象。

- throw关键字用在方法内部，只能用于抛出一种异常，可以跑出代码块和方法中的异常，包括受检异常和非受检异常。
- throws关键字用在方法声明上，可以抛出多个异常。

### 3.5 final，finally，finzlize有什么区别

- final，属于java关键字，可以用来修饰类，字段和方法。类不能被继承，字段不能修改值，方法不能被重写。
- finally一般和try-catch一起使用，通常我们将一定会执行的代码放在finally中，表示不管是否出现异常，该代码块都会执行，一般用来存放一些关闭资源。
- finzlize属于Object方法。在垃圾回收对象之前，该方法会被调用。

### 3.6 NoClassDefFoundError 和 ClassNotFoundException 区别？

NoClassDefFoundError 是一个 Error 类型的异常，是由 JVM 引起的，不应该尝试捕获这个异常。

引起该异常的原因是 JVM 或 ClassLoader 尝试加载某类时在内存中找不到该类的定义，该动作发生在运行期间，即编译时该类存在，但是在运行时却找不到了，可能是变异后被删除了等原因导致；

ClassNotFoundException 是一个受查异常，需要显式地使用 try-catch 对其进行捕获和处理，或在方法签名中用 throws 关键字进行声明。当使用 Class.forName, ClassLoader.loadClass 或 ClassLoader.findSystemClass 动态加载类到内存的时候，通过传入的类路径参数没有找到该类，就会抛出该异常；另一种抛出该异常的可能原因是某个类已经由一个类加载器加载至内存中，另一个加载器又尝试去加载它

### 3.7 try-catch-finally 中哪个部分可以省略？

答：catch 可以省略

### 3.8 try-catch-finally 中，如果 catch 中 return 了，finally 还会执行吗？

答：会执行，在 return 前执行。

### 3.9 常见的RuntimeException有哪些

ClassCastException类转换异常

IndexOutOfBoundsException数组越界异常

NullPointerException空指针

ArrayStoreException数组存储异常

